File: day4/valid_records.rb
#!/usr/bin/env ruby

v1 = 0
v2 = 0

def check_body?(head, body)
  if head == "byr"
    is_v_byr?(body.to_i)
  elsif head == "iyr"
    is_v_iyr?(body.to_i)
  elsif head == "eyr"
    is_v_eyr?(body.to_i)
  elsif head == "hgt"
    is_v_hgt?(body)
  elsif head == "hcl"
    is_v_hcl?(body)
  elsif head == "ecl"
    is_v_ecl?(body)
  elsif head == "pid"
    is_v_pid?(body)
  elsif head == "cid"
    is_v_cid?(body)
  end
end

def is_v_cid?(y)
  true
end

def is_v_byr?(y)
  y >= 1920 && y <= 2002
end

def is_v_iyr?(y)
  y >= 2010 && y <= 2020
end

def is_v_eyr?(y)
  y >= 2020 && y <= 2030
end

def is_v_hgt?(hs)
  if hs.end_with?("cm") 
    h = hs.to_i
    h >= 150 && h <= 193
  elsif hs.end_with?("in")
    h = hs.to_i
    h >= 59 && h <= 76
  else
    false
  end
end

def is_v_hcl?(cl)
  cl.match(/^#[0-9a-f]{6}$/) != nil
end

def is_v_ecl?(cl)
  ["amb", "blu","brn","gry","grn","hzl","oth"].include? cl
end

def is_v_pid?(ps)
  ps.match(/^[0-9]{9}$/) != nil
end

File.readlines('day_4_data.txt').each do |line|
    is_v1 = line.include?("byr:") &&
           line.include?("iyr:") &&
           line.include?("eyr:") &&
           line.include?("hgt:") &&
           line.include?("hcl:") &&
           line.include?("ecl:") &&
           line.include?("pid:")
    if is_v1 
      v1 += 1
      is_v2 = true
      line.split(" ").each do |item|
        head,body = item.split(":")
        if not check_body?(head, body)
          is_v2 = false
          break
        end
      end

      if is_v2 
        v2 += 1
        puts line
      end
    end
end

puts v1
puts v2

File: day4/valid_records.hs
#!/usr/bin/env stack
-- stack --resolver lts-16.19 script
--{-# LANGUAGE OverloadedStrings #-}

-- file contents
--
--
import Data.List
--import Data.Bits
--import qualified Data.Text as T
import Text.Regex.TDFA
import Data.Char

type Entry = [String]
type Answer = (Int, Int)

main = interact $ showResult . results . map parseRaw . lines

showResult :: Answer -> String
showResult (p1, p2) = "Part I answer: " ++ show p1 ++ ".\nPart II answer: " ++ show p2 ++ ".\n"

parseRaw :: String -> Entry
parseRaw = words

results :: [Entry] -> Answer
results es = (length validFieldsEntries,  length $ filter (all checkValeu) validFieldsEntries) 
  where 
    validFieldsEntries = filter (\e -> (length $ deleteBy isPrefixOf "cid:" e) >= 7) es
    checkValeu :: String -> Bool
    checkValeu e = case splitAt 4 e of 
                     ("byr:", vs) -> vs =~ "^[0-9]{4}$" && let yr = read vs in yr >= 1920 && yr <= 2002
                     ("iyr:", vs) -> vs =~ "^[0-9]{4}$" && let yr = read vs in yr >= 2010 && yr <= 2020
                     ("eyr:", vs) -> vs =~ "^[0-9]{4}$" && let yr = read vs in yr >= 2020 && yr <= 2030
                     ("hgt:", vs) -> vs =~ "^([0-9]{2}in|[0-9]{3}cm)$" && if "cm" `isSuffixOf` vs then read (take 3 vs) `elem` [150..193] else read (take 2 vs) `elem` [59..76]
                     ("hcl:", vs) -> vs =~ "^#[0-9a-f]{6}$"
                     ("ecl:", vs) -> vs `elem` ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]
                     ("pid:", vs) -> vs =~ "^[0-9]{9}$"
                     _ -> True

